src/nnet3bin/nnet3-xvector-signal-perturb-egs.cc

```diff
@@ -0,0 +1,155 @@
+  PerturbXvectorSignal perturb_xvector(opts);
+  
+  Matrix<BaseFloat> shifted_egs(input_egs);
+  // Generate random shift samples to shift egs. 
+  if (opts.max_shift != 0.0) {
+    int32 max_shift_int = static_cast<int32>(opts.max_shift * opts.frame_dim);
+    // shift input_egs using random shift. 
+    int32 eg_dim = input_egs.NumCols() - opts.frame_dim,
+      shift = RandInt(0, max_shift_int);
+    shifted_egs.CopyFromMat(input_egs.Range(0, input_egs.NumRows(), shift, eg_dim));
+  }
+  
+  Matrix<BaseFloat> rand_distort_shifted_egs(shifted_egs);
+  if (opts.rand_distort) {
+    // randomly generate an zero-phase FIR filter with no zeros.
+    // In future, we can select trucated part of room impluse response
+    // and convolve it with input_egs.
+    perturb_xvector.ComputeAndApplyRandDistortion(shifted_egs,
+                                  &rand_distort_shifted_egs);
+  }
+
+  if (noise_egs) { 
+    // select random block of noise egs and add to input_egs
+    // number of additive noises should be larger than number of input-egs.
+    KALDI_ASSERT(noise_egs->NumRows() >= input_egs.NumRows());
+    if (noise_egs->NumRows() < input_egs.NumRows()) {
+      // repeat the noise_egs_mat blocks to have same length block
+      // and randomly perturb the rows.
+    } else {
+      // Select random submatrix out of noise_egs and add it to perturb_egs.
+      // we should shuffle noise_egs before passing them to this binary.
+      int32 start_row_ind = RandInt(0, noise_egs->NumRows() - input_egs.NumRows()),
+        start_col_ind = RandInt(0, noise_egs->NumCols() - input_egs.NumCols()); 
+      rand_distort_shifted_egs.AddMat(1.0, noise_egs->Range(start_row_ind, input_egs.NumRows(),
+                                      start_col_ind, input_egs.NumCols()));
+    }
+  }
+  // Perturb speed of signal egs
+  Matrix<BaseFloat> warped_distorted_shifted_egs(rand_distort_shifted_egs);
+  if (opts.max_time_stretch != 0.0) 
+    perturb_xvector.TimeStretch(rand_distort_shifted_egs, 
+                                &warped_distorted_shifted_egs);
+   
+  // If nagation is true, the sample values are randomly negated
+  // with some probability.
+  if (opts.negation) {
+   
+  }
+}
+
+} // end of namespace nnet3
+} // end of namespace kaldi
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    using namespace kaldi::nnet3;
+    typedef kaldi::int32 int32;   
+    typedef kaldi::int64 int64;
+
+
+    const char *usage =
+        "Corrupts  the examples supplied via input pipe with different type of distortions\n"
+        "such as additive noise, negation, random time shifts or random distortion.\n"
+        "Usage: nnet3-xvector-signal-perturb-egs [options...] <egs-especifier> <egs-wspecifier>\n"
+        "e.g.\n"
+        "nnet3-xvector-signal-perturb-egs --noise-egs=noise.egs\n"
+        "--max-shift=0.2 --max-speed-perturb=0.1 --negation=true\n"
+        "ark:input.egs akr:distorted.egs\n";
+    ParseOptions po(usage);
+
```

Comment:
a program like this will need to support a --rand-seed option (or are we calling it srand?).